### PR TITLE
feat: pay fee by real feeRate

### PR DESCRIPTION
**Pull request by zhangyouxin**

### CODE REVIEW

File: src/light-godwoken/LightGodwokenV0.ts [hunk 1]
```diff
@@ -222,29 +227,31 @@ export default class DefaultLightGodwokenV0 extends DefaultLightGodwoken impleme
       args: rollupTypeHash + ethAddress.slice(2),
     };
     const accountScriptHash = utils.computeScriptHash(l2AccountScript);
-    console.log("account script hash:", accountScriptHash);
+    debug("account script hash:", accountScriptHash);
     const fromId = await this.godwokenClient.getAccountIdByScriptHash(accountScriptHash);
     if (!fromId) {
       throw new Error("account not found");
     }
     const isSudt = payload.sudt_script_hash !== "0x0000000000000000000000000000000000000000000000000000000000000000";
```
Comment (homura) [inline src/light-godwoken/LightGodwokenV0.ts:235]: check byte(s) like data with string comparison may cause some bugs, e.g. eth address `0xCcCCccccCCCCcCCCCCCcCcCccCcCCCcCcccccccC` and `0xcccccccccccccccccccccccccccccccccccccccc`

we can compare it with a utility function

File: src/light-godwoken/LightGodwokenV0.ts [hunk 2]
```diff
@@ -387,8 +395,15 @@ export default class DefaultLightGodwokenV0 extends DefaultLightGodwoken impleme
         return cell_deps.push(getCellDep(layer1Config.SCRIPTS.sudt));
       });
     }
-    txSkeleton = await this.injectCapacity(txSkeleton, l1Lock, BigInt(0));
-    const signedTx = await this.provider.signL1Transaction(txSkeleton);
+    txSkeleton = await this.appendPureCkbCell(txSkeleton, l1Lock, BI.from(1000));
```
Comment (sighWang) [inline src/light-godwoken/LightGodwokenV0.ts:398]: why 1000 here?

File: src/light-godwoken/lightGodwoken.ts [hunk 3]
```diff
@@ -170,16 +224,29 @@ export default abstract class DefaultLightGodwoken implements LightGodwokenBase
...
+      // exchange sudt if any left after deposit
+      if (BI.from(sudtData).gt(BI.from(0))) {
+        outputCells = [exchangeCell].concat(...outputCells);
+        exchangeCell.cell_output.capacity = exchangeCapacity.sub(sudtCapacity).toHexString();
+      } else {
+        exchangeCell.cell_output.capacity = `0x${exchangeCapacity.toString(16)}`;
+      }
 
-    return [outputCell, exchangeCell];
+      if (BI.from(exchangeCell.cell_output.capacity).gte(BI.from(6300000000))) {
```
Comment (sighWang) [inline src/light-godwoken/lightGodwoken.ts:238]: what if "else"?

File: src/light-godwoken/lightGodwoken.ts [hunk 4]
```diff
@@ -159,9 +211,11 @@ export default abstract class DefaultLightGodwoken implements LightGodwokenBase
 
     if (payload.sudtType && payload.amount && payload.amount !== "0x" && payload.amount !== "0x0") {
       outputCell.cell_output.type = payload.sudtType;
-      outputCell.data = utils.toBigUInt128LE(BigInt(payload.amount));
+      outputCell.data = utils.toBigUInt128LE(payload.amount);
+      let outputCells = [outputCell];
 
-      const sudtData = utils.toBigUInt128LE(sumSustAmount - BigInt(payload.amount));
+      // contruct sudt exchange cell
+      const sudtData = utils.toBigUInt128LE(sumSustAmount.sub(payload.amount));
```
Comment (sighWang) [inline src/light-godwoken/lightGodwoken.ts:218]: rename to sudtData to exchangeSudtAmount and sumSustAmount to sumSudtAmount

File: src/light-godwoken/lightGodwoken.ts [hunk 5]
```diff
@@ -170,16 +224,29 @@ export default abstract class DefaultLightGodwoken implements LightGodwokenBase
...
+      const sudtCapacity = helpers.minimalCellCapacity(exchangeSudtCell);
       exchangeSudtCell.cell_output.capacity = "0x" + sudtCapacity.toString(16);
 
-      // minus sudt capacity from exchange cell
-      exchangeCell.cell_output.capacity = `0x${(exchangeCapacity - sudtCapacity).toString(16)}`;
-      return [outputCell, exchangeCell, exchangeSudtCell];
-    }
+      // exchange sudt if any left after deposit
+      if (BI.from(sudtData).gt(BI.from(0))) {
+        outputCells = [exchangeCell].concat(...outputCells);
```
Comment (sighWang) [inline src/light-godwoken/lightGodwoken.ts:232]: if we have sudt to exchange, maybe we should concat exchangeSudtCell to outputCells

File: src/light-godwoken/lightGodwoken.ts [hunk 6]
```diff
@@ -105,11 +125,44 @@ export default abstract class DefaultLightGodwoken implements LightGodwokenBase
...
+  }
+
+  getTransactionSizeByTx(tx: Transaction): number {
+    const serializedTx = core.SerializeTransaction(toolkit.normalizers.NormalizeTransaction(tx));
+    // 4 is serialized offset bytesize
+    const size = serializedTx.byteLength + 4;
+    return size;
+  }
+
   generateDepositOutputCell(collectedCells: Cell[], payload: DepositPayload): Cell[] {
```
Comment (sighWang) [inline src/light-godwoken/lightGodwoken.ts:166]: this function have a lot of logic,I think we can try to split them to small functions and add unit test for them.and `withdrawal `and `deposit` function is same

File: src/light-godwoken/LightGodwokenV0.ts (hunk 2 above)
Comment (zhangyouxin) [reply src/light-godwoken/LightGodwokenV0.ts:398]: Only `deposit` and `unlock` could issue a layer 1 tx, both of them would not cost more than 1000 shannon tx fee.

File: src/light-godwoken/lightGodwoken.ts (hunk 3 above)
Comment (zhangyouxin) [reply src/light-godwoken/lightGodwoken.ts:238]: It's supposed to be zero, thus no exchange is needed.

File: src/light-godwoken/lightGodwoken.ts (hunk 5 above)
Comment (zhangyouxin) [reply src/light-godwoken/lightGodwoken.ts:232]: should make sure the last element of the array to be a pure ckb cell

File: src/light-godwoken/lightGodwoken.ts (hunk 6 above)
Comment (zhangyouxin) [reply src/light-godwoken/lightGodwoken.ts:166]: You are right, gonna do some refactor here

File: src/light-godwoken/LightGodwokenV0.ts (hunk 2 above)
Comment (sighWang) [reply src/light-godwoken/LightGodwokenV0.ts:401]: may be we can refactor 1000 to a const. and what if 'fromAddress' do not have 1000 shannon, do we need handle?

Comment (zhangyouxin) [reply src/light-godwoken/LightGodwokenV0.ts:401]: could extract a const, error is handled in `appendPureCkbCell` method

**Pull request merged**